### ISSUE TITLE
AGP-8: Adjustable liquid reserves (with reasonable delay)

### ISSUE THREAD

**Issue by andriytyurnikov** [aragon/governance#16]
### Goals

Implement a model, in which token-holders of organization always have access to their fair share of liquidity reserves thanks to reserves being implemented with Bancor formula.

- Increase value of the issued tokens by providing liquidity
- Allow attracting more capital by decreasing negative perception of "over-funded" projects
- Allow minority investors to exercise their freedom to leave organization with their fair share of reserves
- Increase transparency and accountability of crowdfunding initiatives, by implementing step-by-step reserves-to-budget transfers.
- Allow additional trajectories of governing dynamics

### Description
When tokens issued by organization are backed with liquid reserves, then the whole process of organization formation may be seen as a process of initial reserves formation.

After initial formation such organization may have a vote about transferring part of the reserves into budget of organization for operational needs (next stage of product development for example).

So organization development may be seen as step-by-step growth of investor's trust, and gradual access of organization to reserves of founding community.

However, when token is backed by liquid reserves, minority investors will be capable to leave organization with their fair share of reserves without consent of majority investors to do so.

In this model liquid reserves would represent degree of investors trust in organization's achievements and path chosen.

Organization members would be able to diverge from "general line of the majority" and to form alternative organizations to pursue alternative approaches.

Responsible, accountable and transparent organization would be able to gradually access funding from reserves. For example, crowdfunded organization created for software development project would be able to access additional funding from reserves on each milestone of the project.
 So reserve ratio (decreasing from 1 to whatever limit or maybe even to 0) would reflect minority investor confidence, while giving them freedom to "vote with their feet".

Both initial and current reserve ratio may may be exposed via Token's contract API (and crowdfunding UI).

Obviously, every voting that affects parameters of the reserves is a potential threat to the interests of minority that is loosing the voting. Having reasonable delay (24-72 hours) before changing reserves would give loosing minority time to leave with their fair share of reserves.

### Uncertainties
Such degree of minority's freedom is quite uncommon, and some organization founders may prefer to avoid usage of such feature.

**Comment by shyblugs** [comment]
With some Maybe instead of just a liquidity reserve, add a bonding curve for the issuance of new tokens as a general curation market model. 

Token there could be a set period, say every 3 months where token holders could vote on some funds moving from the pool into a discretionary fund.

**Comment by andriytyurnikov** [comment]
@shyblugs , correct me if I am missing something:
1) Bancor formula is a specific case of bonding curve, which does NOT give opportunity (maybe desired and positive - markets will decide) to early stage investors to leverage their "first mover advantage" over later investors. So we are talking about adjustable bonding curve, where Bancor formula is a specific case. Sure. Markets will decide. Also there is theoretical possibility of "down-round", when later investors recieve *better* bonding curve then early ones.
2) I guess need for reasonable delay is not being disputed here.
3) I see periodic and scheduled votings as orthogonal (and useful) feature, that might deserve separate AGP, but I am not sure it is (yet?) relevant to this AGP, as many teams follow rather rough roadmaps instead of strict schedules (which may be justified in case of iterative design and discovery process)

**Comment by andriytyurnikov** [comment]
@shyblugs thanks again for input, please feel free to share good references to math and philosophy of bonding curves

**Comment by lkngtn** [comment]
Hi @andriytyurnikov thanks for posting this, I'm a huge supporter of this idea! If you have not already seen it I suggest checking out the [1Hive proposal](https://github.com/district0x/district-proposals/issues/3) I submitted to district0x earlier this year. The goal and functionality is pretty much exactly as you described and I would love to hear your thoughts on it. 

With that said I do want to point out that your characterization of the Bancor formula seems to be inaccurate. 

> Bancor formula is a specific case of bonding curve, which does NOT give opportunity (maybe desired and positive - markets will decide) to early stage investors to leverage their "first mover advantage" over later investors. 

The bancor formula keeps a constant reserve ratio but there is definitely an advantage to early stage investors. This is because even though the CRR is constant, the the actual price will increase or decrease to maintain that ratio. Meaning that someone buying in early will be able to sell later at a premium (or potentially at a loss depending on the success or failure of the project). 

There is a special case where if you set CRR to 1, there is no advantage to "getting in early" because the token would be pegged to whatever reserve token is used. However, it probably doesn't make sense to ever use a CRR of 1 because there are simpler and more efficient mechanism for creating a 1:1 peg.

**Comment by andriytyurnikov** [comment]
@lkngtn , thank you so much for showing those links, your District0x proposal of [1Hive](https://github.com/district0x/district-proposals/issues/3) is obviously about very same concept, and @simondlr is already started writing code in [ConsenSys/curationmarkets](https://github.com/ConsenSys/curationmarkets) repository.
Let me also link [Curation Markets Whitepaper](https://docs.google.com/document/d/1VNkBjjGhcZUV9CyC0ccWYbqeOoVKT2maqX0rK3yXB20/edit)

Given strong dependencies between Funding and Governance models I hope this efforts would converge at some point. (Actually I am surprised you didn't cross-posted that proposal here - now we correcting that ;)

* First comment is about Reserve Ratio:
I humbly suggest, that in case of reserve ratio being adjustable by governing community, Reserve Ratio with value 1 might be fairly attractive for early stage communities that just start forming around idea or concept. But make no mistake - I am open to critique. It is just not obvious to me why this suggestion might be wrong or useless in case of adjustable Reserve Ratio.

* Second comment is about "first movers advantage" or "asymmetrical value of contributions":
I guess any way may work, depending on culture/motivation of communities, nature of the project and competing communities present on market in general. But I want to make very simple point: *some* communities on *some* stages may prefer to NOT exercise their advantage over those who are invited to join later.

* Third comment is about Bancor formula:
I have no problem with being wrong and learning something in the process. But I am fairly skeptical about existence of "first mover advantage" in case of Reserve Ratio = 1.
And sure I am interested to see how significant is this advantage (still, if it is there) in case of other ReserveRatio value (see [Bancor examples](https://docs.google.com/spreadsheets/d/10Xn--j8W__rMLrkcXxZC66jH2fNXYQAwenqjFZXjUik/edit#gid=877776767)).
Again we should remember that "Bancor formula" might be quite different from whatever other "Bonding curve" you or @shyblugs might have in mind.

* Fourth comment on my motivation and rationale behind this AGP:
From my standpoint current ICO landscape is a perfect example of escalating levels of moral hazard. Community participants with evil motivation have proven themselves to be fast and effective in innovation and copying traits and techniques of non-evil actors. Current ICO model relies on trust to fundraiser persona, and human capacity to trust each other is being exploited heavily because of significant levels of interest to the Crypto space.
Lack of simple and powerful trust-less fundraising and governance infrastructure leads us to the point of exhausted trust.
Having crusade against centralized ICOs or "Early investors bonuses" is not my goal, but I am convinced that alternative is needed.

In general it looks like our discussion goes into stage where specific formulas and scenarios should be used, further discussion and detailed description of formulas are sure needed and welcome.

**Comment by shyblugs** [comment]
@andriytyurnikov yes so if your trying to eliminate the first mover advantage then that type of bonding curve isn't appropriate. However in a real world use case, by charging the same price discounts the risk of early investors. It seems like everyone would be incentivized to wait till the last possible moment to contribute

**Comment by andriytyurnikov** [comment]
> Wouldn't it make sence to distribute tokens weighted by contribution?

@shyblugs I believe Bancor formula does that - dynamically.
However, after your valuable input, I am totally open to having adjustable bonding curve, which may or may not be applied and changed by communities.

**Comment by shyblugs** [comment]
This is a really good idea. What I'd like to see is a few Refrence models to point to for others interested in building on top of AragonOS

**Comment by andriytyurnikov** [comment]
@shyblugs regarding

> However in a real world use case, by charging the same price discounts the risk of early investors. It seems like everyone would be incentivized to wait till the last possible moment to contribute.

I would like to comment that with liquid Bancor-backed reserves there is no initiative whatsoever to delay contribution -  as such contribution may be converted back to reserve token without loss (sure with assumption of reserve ratio not being changed between contribution and liquidation).
Please correct me if I am wrong here

**Comment by shyblugs** [comment]
@andriytyurnikov

So when is the isn't 1:1 the tokens are more expensive to mint as the circulating  supply increases. The assets backing each token are the same at the point of withdrawal

As for tge benefit if investing later. Tge optimal strategy would be to invest as late as possible for 2 reasons 

1. The later you invest the more information you have avalable
2. The net present value of the money is a function of time. In crypto that is alot more time sensitive

**Comment by andriytyurnikov** [comment]
@shyblugs, 

> 1. The later you invest the more information you have available

Well, as withdrawal is lossless, those who had "less information" as you put it - have no disadvantage compared to "late, well informed guys". So it is still challenging for me to understand what is the exploitable advantage *exactly*?

> 2. The net present value of the money is a function of time. In crypto that is alot more time sensitive

Well, volatility-related issues deserve separate explicit discussion, and volatility aside, I am sorry, but what is the exact *exploitable advantage* of late contribution over early contribution, when we are talking about liquid tokens?
